### ISSUE TITLE
[css-overflow-4] Should block-ellipsis clip ::before/::after text?

### ISSUE THREAD

**Issue by fantasai** [w3c/csswg-drafts#11253]
One of the use cases for line-clamp etc. is truncating blockquotes, but there's a problem with the end quotation mark getting dropped when the text is ellipsized.

Should we exempt ::before/::after from trunctation and place it after the ellipsis?

_Related issue for `text-ellipsis`: https://github.com/w3c/csswg-drafts/issues/9541 That one's trickier because you'd need to be tracking the stack of opening/closing punctuation._

**Comment by yisibl** [comment]
Yes, it might help us to attach a button style feature after the ellipsis.
This button is a very common pattern in actual web pages or components.

![image](https://github.com/user-attachments/assets/71a74bd8-7421-42ed-813d-01613c6d40cf)

See https://vue-clamp.vercel.app

**Comment by Loirooriol** [comment]
What if the `::after` has lots of contents that would increase the number of lines?

I'm also not sure if it makes sense to always do this for `::after`, and I could see authors wanting this behavior for a real element (e.g. in @yisibl the button should probably be an actual `<button>` so that it's keyboard navigable and such).

So should we have a property to control this?

**Comment by andreubotella** [comment]
> Yes, it might help us to attach a button style feature after the ellipsis. This button is a very common pattern in actual web pages or components.

In the example you show, this is not done using `(-webkit-)line-clamp`, but instead the text actually has an ellipsis. I think most uses of `-webkit-line-clamp` that do something like this (with the toggle button inside the line-clamp container, rather than in a parent) do so with `position: absolute`, and we recently resolved that that behavior should work if the abspos's containing block is the element with `-webkit-line-clamp` or an ancestor (#11379).

That said, that wouldn't make it possible to have the toggle button inline with the text. So there's a use case to enable this. However, I think this would bring up a number of tricky issues to solve. It would also be best suited to the `continue: discard` version of line-clamp, rather than the `continue: collapse` variant I'm working on in the Chrome implementation (see #7708).

**Comment by yisibl** [comment]
Another very common use case is that we usually want to display the ellipsis only when the text inside a Chinese book title exceeds the length. I think this also applies to other punctuation marks that appear in pairs.

```html
data:text/html;charset=UTF-8,<!DOCTYPE html>
<style>
    h2::before {
      content: "《";
    }
    h2::after {
        content: "》";
    }
    h2 {
        white-space: nowrap;
        overflow: hidden;
        width: 12em;
        text-overflow: ellipsis;
        text-overflow: "...》";
    }
</style>
<h2>编写可维护的JavaScript</h2>
```

Currently, we can do this with `text-overflow: “...” ` to achieve this effect indirectly, but this will lead to inaccurate width calculation. The fundamental solution is to exclude the clip `::before/after`.

Firefox:

![Image](https://github.com/user-attachments/assets/643f774b-7c0f-419a-9f9b-80d9cb5641d9)

**Comment by frivoal** [comment]
For the button use case, I agree with @andreubotella that there are other patterns through which this is usually achieved.

For the purpose of closing quotation marks, I think the pattern shown in https://github.com/w3c/csswg-drafts/issues/11253#issuecomment-2821606428 shows how this is solvable without special-casing `::after`. Adjusting it for `block ellipsis`, it would look something like this:

```css
blockquote::before { content: "《"; }
blockquote::after { content: "》"; }
blockquote {
  line-clamp: 2 "…》";
}
```

This is quite manageable. Sure, it requires some cooperation between what sets the closing quote and what sets the ellipsis, but you'll need some anyway to even know that the `::after` is being used for a closing quote, so that seem fine.

Arguably, one extension we might want, to build up on that pattern, but in a more generic way, is to allow this as a value of `block-ellipsis` instead of just `<string>`:
> [ \<string\> | [\<quote\>](https://www.w3.org/TR/css-content-3/#typedef-quote) ]+

Though the behavior should be different from when it's used in the `content` property, and not change the level of nesting of the quotes, since that would be redundant with `::after`, which is hidden, but still applied.

That said, while we can expand in that direction, I'm not sure I'd rush there before there's clearly established demand for this extra complexity. We can always come back to it later. So my proposal would be:
* don't special case `::after`
* defer to a later level the ability to use `close-quote`
